HouseRules_Essentials/Rules/AbilityRandomPieceListRule.cs

```diff
@@ -0,0 +1,54 @@
+﻿namespace HouseRules.Essentials.Rules
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using Boardgame;
+    using Boardgame.BoardEntities.Abilities;
+    using DataKeys;
+    using HarmonyLib;
+    using HouseRules.Types;
+    using UnityEngine;
+
+    public sealed class AbilityRandomPieceListRule : Rule, IConfigWritable<Dictionary<string, BoardPieceId[]>>
+    {
+        public override string Description => "Ability randomPieceLists are replaced";
+
+        private readonly Dictionary<string, BoardPieceId[]> _adjustments;
+        private readonly Dictionary<string, BoardPieceId[]> _originals;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AbilityRandomPieceListRule"/> class.
+        /// </summary>
+        /// <param name="adjustments">Dict of ability name string and list of BoardPieceID[]s
+        /// Replaces the list of pieces that certain abilities will spawn (e.g. NaturesCall, SpawnCultists etc).</param>
+        public AbilityRandomPieceListRule(Dictionary<string, BoardPieceId[]> adjustments)
+        {
+            _adjustments = adjustments;
+        }
+
+        public Dictionary<string, BoardPieceId[]> GetConfigObject() => _adjustments;
+
+        protected override void OnPostGameCreated(GameContext gameContext)
+        {
+            var abilities = Resources.FindObjectsOfTypeAll<Ability>();
+            foreach (var item in _adjustments)
+            {
+                var ability = abilities.First(c => c.name.Equals($"{item.Key}(Clone)"));
+                _originals[item.Key] = ability.randomPieceList;
```

Comment:
Is an error logged when this rule's `OnPostGameCreated()` is called by core?

`_originals` is never instantiated - I would think we need `_originals = new Dictionary<string, BoardPieceId[]>();` in the constructor.

Reply:
No errors in the logs for activiating rules, but I get one for 'Deactivating ruleset'

```
[20:23:26.215] [HouseRules:Core] [WARNING] Failed to deactivate rule [HouseRules.Essentials.Rules.AbilityRandomPieceListRule]: System.NullReferenceException: Object reference not set to an instance of an object

  at HouseRules.Essentials.Rules.AbilityRandomPieceListRule.OnDeactivate (Boardgame.GameContext gameContext) [0x00008] in <f409b40973284dd5939c47abf2059d57>:0 

  at HouseRules.HR.TriggerDeactivateRuleset (Boardgame.GameContext gameContext) [0x00084] in <237c0339bf86440a997f7e8b84fd392b>:0 
```

Time for me to try and work out what a 'constructor' is I guess.

Reply:
Force pushed a fresh commit with the fix in place. No errors in the melonloader logs anymore.